### PR TITLE
Core: Show exception rather than error on react error boundary

### PR DESCRIPTION
**Pull request by shilman**
Issue: #8087 

## What I did

(Follow-up to #8097)

Use `showExecption` (story exception) rather than `showError` (invalid object returned from story function) on React Error Boundary.

## How to test

- official-storybook
- chromatic (on this PR)

### CODE REVIEW

**Review by tmeasday — APPROVED** [review]
This seems correct to me but would be interested in hearing from @Hypnosphi.

If the user code throws we want to do `storyThrewException` with a proper error object.

**Comment by Hypnosphi** [comment]
Can you please take a screenshot of example error? Does it contain duplicated stacktrace?

**Comment by Hypnosphi** [comment]
> chromatic (on this PR)

Not really, see https://github.com/storybookjs/storybook/issues/8087#issuecomment-532029161

**Comment by shilman** [comment]
@Hypnosphi 

Storybook

File: app/react/src/client/preview/render.tsx [hunk 1]
```diff
@@ -32,10 +32,10 @@ class ErrorBoundary extends React.Component<{
     }
   }
 
-  componentDidCatch({ message, stack }: Error) {
-    const { showError } = this.props;
+  componentDidCatch(err: Error) {
+    const { showException } = this.props;
     // message partially duplicates stack, strip it
```
Comment (Hypnosphi) [inline app/react/src/client/preview/render.tsx]: Let's actually do it again

**Comment by shilman** [comment]
Storybook

File: app/react/src/client/preview/render.tsx (hunk 1 above)
Comment (shilman) [reply app/react/src/client/preview/render.tsx]: Hmm will wait for @tmeasday 's feedback on this. Apparently it interacts with Chromatic in some way

Comment (Hypnosphi) [reply app/react/src/client/preview/render.tsx]: chromatic can't test pages with errors anyway

**Comment by tmeasday** [comment]
Can we distinguish the two types of errors in the `componentDidCatch`? Theoretically the first case (user error) should be passed unmodified to `storyThrewException` and the second as an object to `storyErrored` (confusing names, I know).

**Comment by ndelangen** [comment]
What's blocking this?

**Comment by shilman** [comment]
> Can we distinguish the two types of errors in the componentDidCatch?

I don't know how to do this

File: app/react/src/client/preview/render.tsx (hunk 1 above)
Comment (Hypnosphi) [reply app/react/src/client/preview/render.tsx]: Please at least remove the comment if we're not actually stripping anything

Comment (tmeasday) [reply app/react/src/client/preview/render.tsx]: @Hypnosphi FYI the issue here was that tools like Chromatic listen to to the `storyThrewException` event in order to show the user a stack trace of broken stories. This was reverting a regression where the wrong event was fired in that case.

Would be good to handle the distinction between "user code error" and "react error" somehow though.

**Pull request merged**